### DISCUSSION TITLE
How to conditionally select a recipe?

### DISCUSSION THREAD

**Discussion by donatello** [casey/just#2315] [Q&A]
I would like to conditionally select a recipe based on argument. For example, I have a list of projects in golang, and a list of projects in rust. I want to build any one of them based on the name of the project.

I want to execute something like:

```
$ just build g1 # A go project
OR
$ just build r1 # A rust project.
```
and have just pick the right recipe. Here is what I tried - a recipe called `build` that selects the recipe to run (as a dependency) based on the name of the project:

```
export go_projects := "g1 g2 g3"
export rust_projects := "r1 r2"

build $project: if `echo $go_projects | grep -q --line-regexp $project && echo go` == "go"` {
  (build_go project)
} else if `echo $rust_projects | grep -q --line-regexp $project && echo rust` == "rust"` {
  (build_rust project)
} else {
  (bad_project_msg project)
}

build_rust $project:
    echo "Building rust project $project"

build_go $project:
    echo "Building go project $project"

bad_project_msg $project:
     echo "unknown project $project"
```

Obviously this doesn't work because specifying this type of conditional dependency for `build` is not supported. Is there an recommended way to achieve this, ideally without recursively calling just?

**Comment by wellkilo** [comment]
With current just (1.53+), this can be expressed without recursively invoking just by using list-valued conditional dependencies. Lists are currently unstable, so enable them explicitly:

    set unstable
    set lists

    go_projects := ['g1', 'g2', 'g3']
    rust_projects := ['r1', 'r2']

    build project:
      #!/usr/bin/env sh
      case '{{project}}' in
        g1|g2|g3) exec go build ./cmd/'{{project}}' ;;
        r1|r2)    exec cargo build --manifest-path '{{project}}/Cargo.toml' ;;
        *)        echo 'unknown project: {{project}}' >&2; exit 2 ;;
      esac

For ordinary builds, dispatching inside one recipe is still the clearest solution and avoids a child just process.

If build_go/build_rust must remain recipes, current list semantics let a false conditional yield [] and a mapped dependency run zero times. For example:

    set unstable
    set lists

    build project: *(build_go *(if project =~ ['^g[1-3]$'] { [project] })) *(build_rust *(if project =~ ['^r[1-2]$'] { [project] }))

    build_go project:
        echo 'Building go project {{project}}'

    build_rust project:
        echo 'Building rust project {{project}}'

Add an assert/error guard for unknown names if you use that form. The relevant behavior is documented in [Lists](https://github.com/casey/just/blob/5969e396f4e5c98b64a5df929e3ee67629295d6e/README.md#lists): a missing else evaluates to [], and mapped dependencies run once per list element. I also verified the shown dependency syntax against current just 1.58.0. On older just versions, use the single shell/shebang recipe above because conditional dependencies were not supported.
